src/cpp/commandreply.cpp

```diff
@@ -30,236 +30,294 @@
-      }
-      break;
-    case REDIS_REPLY_INTEGER:
-      std::cout<<index_tracker + " value: "
-               <<this->_reply->integer<<std::endl;
-      break;
-    case REDIS_REPLY_DOUBLE:
-      std::cout<<index_tracker + " value: "
-               <<this->_reply->dval<<std::endl;
-      break;
-    case REDIS_REPLY_ERROR:
-      std::cout<<index_tracker + " value: "
-               <<std::string(this->str(), this->str_len())
-               <<std::endl;
-      break;
-    case REDIS_REPLY_BOOL:
-      std::cout<<index_tracker + " value: "
-               <<this->_reply->integer<<std::endl;
-      break;
-    default:
-      std::cout<<index_tracker
-               <<" value type not supported."<<std::endl;
-  }
-  return;
+    //TODO these recursive functions can't use 'this' unless
+    //we have a constructor that takes redisReply*
+    std::cout<<index_tracker + " type: "
+             <<redis_reply_type()<<std::endl;
+    switch (_reply->type) {
+        case REDIS_REPLY_STRING:
+            std::cout<<index_tracker + " value: "
+                     <<std::string(str(),
+                                   str_len())
+                     <<std::endl;
+            break;
+        case REDIS_REPLY_ARRAY:
+            for (size_t i = 0; i < n_elements(); i++) {
+              std::string r_prefix = index_tracker + "[" +
+                                     std::to_string(i) + "]";
+              CommandReply tmp = (*this)[i];
+              tmp.print_reply_structure(r_prefix);
+            }
+            break;
+        case REDIS_REPLY_INTEGER:
+            std::cout<<index_tracker + " value: "
+                     <<_reply->integer<<std::endl;
+            break;
+        case REDIS_REPLY_DOUBLE:
+            std::cout<<index_tracker + " value: "
+                     <<_reply->dval<<std::endl;
+            break;
+        case REDIS_REPLY_ERROR:
+            std::cout<<index_tracker + " value: "
+                     <<std::string(str(), str_len())
+                     <<std::endl;
+            break;
+        case REDIS_REPLY_BOOL:
+            std::cout<<index_tracker + " value: "
+                     <<_reply->integer<<std::endl;
+            break;
+        default:
+            std::cout<<index_tracker
+                     <<" value type not supported."<<std::endl;
+    }
+}
+
+redisReply* CommandReply::deep_clone_reply(const redisReply* reply)
+{
+    if (reply == NULL)
+        return NULL;
+    redisReply* redis_reply = new redisReply;
+    *redis_reply = *reply;
+    redis_reply->str = NULL;
+    redis_reply->element = NULL;
+    switch (redis_reply->type) {
+        case REDIS_REPLY_ARRAY:
+        case REDIS_REPLY_MAP:
+        case REDIS_REPLY_SET:
+            // allocate memory for element and do deep copy
+            if (redis_reply->elements > 0) {
+                redis_reply->element = new redisReply*[redis_reply->elements];
+                if (reply->element != NULL) {
```

Comment:
This is a good catch to make sure we don't do ``reply->element[i]`` if ``reply`` is ``NULL``.  However, I wonder if we should have an else case?  

That is, if ``reply->element`` is NULL, that means ``reply->elements`` and ``reply->element`` are in inconsistent states.  If that is true, we will have ``redis_repy->element`` set to an array of ``redis_reply*`` structs with default values.  This would be inconsistent with the input ``reply`` object.  We should either free memory and set ``redis_reply->element`` to ``NULL`` or throw an error because the input reply is invalid.  What do you think?